streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java

```diff
@@ -416,26 +552,43 @@ private long currentSegmentLastTime() {
         }
 
         private void getNextSegmentIterator() {
-            ++currentSegmentId;
-            lastSegmentId = cacheFunction.segmentId(Math.min(timeTo, maxObservedTimestamp.get()));
+            if (forward) {
+                ++currentSegmentId;
+                lastSegmentId = cacheFunction.segmentId(Math.min(timeTo, maxObservedTimestamp.get()));
 
-            if (currentSegmentId > lastSegmentId) {
-                current = null;
-                return;
-            }
+                if (currentSegmentId > lastSegmentId) {
+                    current = null;
+                    return;
+                }
 
-            setCacheKeyRange(currentSegmentBeginTime(), currentSegmentLastTime());
+                setCacheKeyRange(currentSegmentBeginTime(), currentSegmentLastTime());
 
-            current.close();
-            current = context.cache().range(cacheName, cacheKeyFrom, cacheKeyTo);
+                current.close();
+
+                current = context.cache().range(cacheName, cacheKeyFrom, cacheKeyTo);
+            } else {
+                --currentSegmentId;
+//                lastSegmentId = cacheFunction.segmentId(Math.min(timeTo, maxObservedTimestamp.get()));
```

Comment:
Guessing this is not actually meant to be commented out

Reply:
actually it could be removed: `lastSegmentId` should be stable going backwards based on `timeFrom`.

On the `if` branch, I assume that `lastSegmentId` could change in between iterations if `maxObservedTimestamp` is updated right?

Reply:
Ah, good catch. That sounds right, the ending segment shouldn't change when you iterate backwards